numpyro/infer/elbo.py

```diff
@@ -73,6 +80,104 @@ def single_particle_elbo(rng_key):
             return - jnp.mean(vmap(single_particle_elbo)(rng_keys))
 
 
+def _get_log_prob_sum(site):
+    if site['intermediates']:
+        log_prob = site['fn'].log_prob(site['value'], site['intermediates'])
+    else:
+        log_prob = site['fn'].log_prob(site['value'])
+    log_prob = scale_and_mask(log_prob, site['scale'])
+    return jnp.sum(log_prob)
+
+
+def _check_mean_field_requirement(model_trace, guide_trace):
+    """
+    Checks that the guide and model sample sites are ordered identically.
+    This is sufficient but not necessary for correctness.
+    """
+    model_sites = [name for name, site in model_trace.nodes.items()
+                   if site["type"] == "sample" and name in guide_trace.nodes]
+    guide_sites = [name for name, site in guide_trace.nodes.items()
+                   if site["type"] == "sample" and name in model_trace.nodes]
+    assert set(model_sites) == set(guide_sites)
+    if model_sites != guide_sites:
+        warnings.warn("Failed to verify mean field restriction on the guide. "
+                      "To eliminate this warning, ensure model and guide sites "
+                      "occur in the same order.\n" +
+                      "Model sites:\n  " + "\n  ".join(model_sites) +
+                      "Guide sites:\n  " + "\n  ".join(guide_sites))
+
+
+class MeanFieldELBO(ELBO):
```

Comment:
@fritzo @neerajprad I need a bit of discussion for this. In NumPyro, `Trace_ELBO` -> `ELBO`, then how about `TraceMeanField_ELBO`? Do we have a plan to rename Pyro objectives for Pyro 2.0?

Reply:
In the one hand, I generally feel that users can tolerate a very limited amount of breakage before wanting to switch to another library, and I'd prefer to spend our limited breakage budget on interesting structural changes (like `numpyro.scan` and `PyroModule`) rather than on boring renamings. OTOH if people really think it's worth renaming, we could do so in Pyro 2 (but still provide backwards-compatible aliases).

Reply:
If we don't have plan to rename in Pyro, then it is better to use TraceMeanField_ELBO here in my opinion. WDYT @neerajprad ?

Reply:
I think the reason why it was `Elbo` in numpyro is because the initial interface was based off of minipyro (rather than Pyro), which had dropped the `Trace` prefix. If we add the `Trace` prefix, we should do so uniformly. It will be more confusing to see `Elbo` and `TraceMeanFieldELBO`. In any case, the generic API should still use the names that we are using in Pyro, and resolve other interface differences which already exist.

p.s. - I thought we started with `Elbo` as the class name and not `ELBO`. The former would be the preferred way of naming classes that are acronyms as per pep-8.

Reply:
Thanks for the input, @neerajprad! It seems that I renamed `Elbo` to `ELBO` at some point in the past...

> If we add the Trace prefix, we should do so uniformly.

Which one do you prefer? My preference order is (top means "like most"):
+ Trace_ELBO and TraceMeanField_ELBO as in Pyro, deprecate ELBO. This is good because we have matched much of the code in early development to Pyro.
+ ELBO and TraceMeanField_ELBO as in this PR. This is also good to me and I don't see the confusion here for users.
+ ELBO and MeanFieldELBO as in the early stage of this PR
+ ELBO and MeanField_ELBO ...
+ Other options?

Reply:
I think (1) sounds most reasonable, simply because the names match exactly with Pyro which dwarfs any other minor considerations. If we do not want to add deprecation warnings in this release, we can go with (2) and change `ELBO` to `Trace_ELBO` in a later release. However, since you have already made other changes to match Pyro API, we might as well bundle these changes with this release (maybe this should be a `0.4` release instead of a patch release?).

Reply:
Good idea! Thanks, I'll go with that. I feel having 0.4 release is more aligned with JAX 0.2 release. :D we also have new stuffs rather than just patches.